File: Find the first or last occurrence of a given number in a sorted array/FindFirstLastOccurence.java
//https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
//time complexity: O(logn)
//space complexity: O(1)
class FindFirstLastOccurrence {
    public int[] searchRange(int[] nums, int target) {
        int firstIndex = findIndex(nums, target, true);
        int lastIndex = findIndex(nums, target, false);
        return new int[] { firstIndex, lastIndex };
    }

    public int findIndex(int[] nums, int target, boolean findFirstIndex) {
        int high = nums.length - 1;
        int low = 0;
        int result = -1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            int currNum = nums[mid];
            if (currNum > target) {
                high = mid - 1;
            } else if (currNum < target) {
                low = mid + 1;
            } else if (currNum == target && findFirstIndex) {
                high = mid - 1;
                result = mid;
            } else if (currNum == target && !findFirstIndex) {
                low = mid + 1;
                result = mid;
            }
        }
        return result;
    }
}

File: Find the first or last occurrence of a given number in a sorted array/FindFirstLastOccurence.py
#https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
#time complexity: O(logn)
#space complexity: O(1)
class Solution:

    def count(self,arr, n, x):
        # code here
        first_index = self.findIndex(arr,n,x,True)
        second_index = self.findIndex(arr,n,x,False)
        if first_index == -1 and second_index == -1:
            return 0
        return second_index - first_index + 1
    
    def findIndex(self,arr,n,x,find_first_index):
        low = 0
        high = len(arr) - 1
        result = -1
        while low <= high:
            mid = (int)(low + (high-low)/2)
            curr_num = arr[mid]
            if curr_num > x:
                high = mid -1
            elif curr_num < x:
                low = mid + 1
            elif curr_num == x and find_first_index:
                high = mid - 1
                result = mid
            elif curr_num == x and not find_first_index:
                low = mid + 1
                result = mid
        return result
    